src/hooks/useInitializeClient.ts

```diff
@@ -19,33 +19,50 @@ along with the library. If not, see <http://www.gnu.org/licenses/>.
 import { useCallback, useEffect } from 'react'
 
 import client from '../api/client'
-import { networkStatusChanged } from '../store/networkSlice'
+import { selectAllAddresses, syncAddressesData } from '../store/addressesSlice'
+import { apiClientInitFailed, apiClientInitSucceeded } from '../store/networkSlice'
 import { useAppDispatch, useAppSelector } from './redux'
 import useInterval from './useInterval'
 
 const useInitializeClient = () => {
   const dispatch = useAppDispatch()
   const network = useAppSelector((state) => state.network)
+  const addressesStatus = useAppSelector((state) => state.addresses.status)
+  const addresses = useAppSelector(selectAllAddresses)
 
   const initializeClient = useCallback(async () => {
     try {
-      await client.init(network.settings)
-      dispatch(networkStatusChanged('online'))
+      await client.init(network.settings.nodeHost, network.settings.explorerApiHost)
+      dispatch(apiClientInitSucceeded())
       console.log(`Client initialized. Current network: ${network.name}`)
     } catch (e) {
+      dispatch(apiClientInitFailed())
       console.error('Could not connect to network: ', network.name)
       console.error(e)
-      dispatch(networkStatusChanged('offline'))
     }
-  }, [dispatch, network.name, network.settings])
+  }, [network.settings.nodeHost, network.settings.explorerApiHost, network.name, dispatch])
 
+  // Is there a better way to trigger the initial client initialization?
+  // Currently we trigger it "magically" by setting the the networkSlice status status to 'connecting', which is
+  // happening when loading the stored network settings and when network settings are updated by the user.
```

Comment:
> I'd say that initializing the client should be the first thing done once the app is up and renaming - wether the network is up or not.
If the network isn't working for whatever reason, an error should be dispatched accordingly from the initialization function.

So far, so good, we got this already.

> Thinking out loud : in order to decouple the client initialization from state changes, maybe we shouldn't use a hook for this. We should be more explicit and dispatch an initialization action from the places that should trigger the initialization process, instead of listening for state changes from the hook...

This is exactly what I wanted to do and I spent a few hours on it, without finding a proper solution. I was working on the form that edits the network settings. When the user clicks "Save custom network settings", I want to trigger a new network initialization and if it's successful, I want to fetch the data for each address. This is easy, I can just run this code right after the network settings are saved, wait for it to succeed, and then fetch the address data:

```ts
const initializeClient = async () => {
  try {
    await client.init(network.settings.nodeHost, network.settings.explorerApiHost)
    dispatch(apiClientInitSucceeded())
  } catch (e) {
    dispatch(apiClientInitFailed())
  }
}
```

Here's the problem though: One of our requirements is that the app SHOULD keep trying to initialize the client in case the client initialization has failed. I have not found a way to do that AND also remember that the app should fetch the address once it succeeds. I tried using callbacks, but I was not satisfied with my solution because it doesn't work for every case. The only way that works as it should is using 2 `useEffect`s that listen to state changes: one that keeps trying to initialize the client if the network is offline, and one that initializes the address data if their status is "uninitialized" and the network is "online".

> App has started => Run the client initialisation.
If the client isn't available, call the init function again from the network status change listener.

That's what we do at the moment with:

```tsx
useEffect(() => {
  if (network.status === 'connecting') {
    initializeClient()
  }
}, [initializeClient, network.status])

const shouldInitialize = network.status === 'offline'
useInterval(initializeClient, 2000, !shouldInitialize)
```

> And for clarity sake, it would maybe be beneficial to decouple the network state from the client state.
So creating a new "Client" slice may help?

Can't see on top of my head how this could help...

Reply:
NOTE: Following our discussion, we will in the future investigate the potential use of the RTK Query that may allow us to better manager retries and polling.